### PR TITLE
#510 Adding experimental SPI for letting AST-modifying annotation pro…

### PR DESCRIPTION
**Pull request by gunnarmorling**
…cessors tell us about future modifications

An experiment of an SPI towards Lombok

### CODE REVIEW

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java [hunk 1]
```diff
@@ -102,6 +110,20 @@ public TypeFactory(Elements elementUtils, Types typeUtils) {
         implementationTypes.put( ConcurrentNavigableMap.class.getName(), getType( ConcurrentSkipListMap.class ) );
     }
 
+    private static List<AstModifyingAnnotationProcessor> getAstModifyingAnnotationProcessors() {
+        List<AstModifyingAnnotationProcessor> processors = new ArrayList<AstModifyingAnnotationProcessor>();
+
+        ServiceLoader<AstModifyingAnnotationProcessor> loader = ServiceLoader.load(
+                AstModifyingAnnotationProcessor.class
```
Comment (rzwitserloot) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: Make this:

    ServiceLoader.load(AstModifyingAnnotationProcessor.class, AstModifyingAnnotationProcessor.class.getClassLoader());

Without the second param, you get the thread context loader and that's not the right one (all the APs share a special loader and we need that one).

**Comment by rzwitserloot** [comment]
You need to fix one tiny thing in this change and it'll work together with the latest lombok edge release.

The thing to fix is commented in the change (https://github.com/mapstruct/mapstruct/pull/1042/commits/60823f0797efd3b985e1b3e781ad60943c5192fa) and the latest lombok edge release that works with this is at https://projectlombok.org/download-edge.html

**Comment by agudian** [comment]
Nice, we'll try it out!

**Comment by gunnarmorling** [comment]
Hey @rzwitserloot indeed it works if Lombok runs first and also - with the adjustment you suggested - when MapStruct comes first. Very nice!

I'm a tad concerned about your implementation of `AstModifyingAnnotationProcessor`, though. It ignores the type passed in and will return true as soon as `init()` of the processor has been called once.

It may pose a problem in cases where types which are enhanced by Lombok are only generated by another AP. Admittedly, that looks like a far-fetched case (why not generating that type completely then using that other AP?), but it could lead to the case where Lombok has been initialized (so you'll return `true` from `isTypeComplete()`) but not enhanced a given type.

**Comment by gunnarmorling** [comment]
@agudian et al., I've pushed an update. It's good to go from side now.

**Comment by gunnarmorling** [comment]
Pushed another commit to address #1045.

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java [hunk 2]
```diff
@@ -67,6 +70,7 @@
 
     private final Elements elementUtils;
     private final Types typeUtils;
+    private RoundContext roundContext;
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: Make it final as well?

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java [hunk 3]
```diff
@@ -565,4 +569,31 @@ static String trimSimpleClassName(String className) {
         return trimmedClassName;
     }
 
+    private boolean isCleared(TypeMirror type) {
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: Add a little javadoc to explain what specifically "cleared" means for the type?

File: processor/src/main/java/org/mapstruct/ap/spi/AstModifyingAnnotationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+public interface AstModifyingAnnotationProcessor {
+
+    /**
+     * Whether the specified type has been fully processed by this processor or not (i.e. this processor will amend the
+     * given type's structure after this invocation).
+     *
+     * @param type The type of interest
+     * @return {@code true} if this processor has fully processed the given type, {@code false} otherwise.
+     */
+    boolean isTypeComplete(TypeMirror type);
```
Comment (agudian) [inline processor/src/main/java/org/mapstruct/ap/spi/AstModifyingAnnotationProcessor.java]: Also needs to return `true` if the processor has no buisness with that type, right?

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java (hunk 3 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: I think the name isCleared has a double connotation, right? IIUYC it means: "safe to use" or "ready to be used". There's also a connotation: erased.

File: processor/src/main/java/org/mapstruct/ap/spi/AstModifyingAnnotationProcessor.java (hunk 4 above)
Comment (sjaakd) [inline processor/src/main/java/org/mapstruct/ap/spi/AstModifyingAnnotationProcessor.java]: I perhaps miss something in the integration tests.. But I assumed that this needs to be somewhere implemented via SPI, right? I can't  find an implementation in the integration tests. And still it works?

Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/spi/AstModifyingAnnotationProcessor.java]: @sjaakd there is currently no integration test. We could add one once this feature is implemented on the Lombok side. I guess I could add a custom processor somehow imitating what Lombok does, but it doesn't seem worth the trouble.

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java (hunk 2 above)
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: Ok.

File: processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java (hunk 3 above)
Comment (gunnarmorling) [inline processor/src/main/java/org/mapstruct/ap/internal/model/common/TypeFactory.java]: @sjaakd, I see where you are coming from. Happy to change it, any suggestions?

**Comment by gunnarmorling** [comment]
@agudian @sjaakd Rebased and added one more commit for addressing your review comments. Good to go fro my side. Thanks for reviewing!

**Comment by sjaakd** [comment]
all ok from my side.

**Comment by gunnarmorling** [comment]
Squashed and merged.

**Comment by kevinmic** [comment]
Were these changes ever merged into the master branch?    There is the comment "Squashed and merged" but when I checkout the source code I can't find the changes.

**Pull request closed without merging**